docs/user/reference/hook.md

```diff
@@ -137,6 +142,236 @@ The result is that the events are consumed in the order: `2-1-3`. Beware.
+
+These hooks require an associated workload/container.
+The hook names that these kinds represent will be prefixed by the workload/container name; for example, `mycontainer-pebble-ready`.
+
+The following environment variables may be set depending on the exact hook:
+ 
+* $JUJU_WORKLOAD_NAME holds the name of the container to which the hook pertains.
+* $JUJU_NOTICE_ID holds the Pebble notice ID.
+* $JUJU_NOTICE_TYPE holds the Pebble notice type.
+* $JUJU_NOTICE_KEY holds the Pebble notice key.
+* $JUJU_PEBBLE_CHECK_NAME holds the name of the Pebble check.
+
+## Hook ordering
+
+A charm's lifecycle consists of distinct **phases**:
+* install
+* operation
+* upgrade
+* tear down
+
+Generally, no assumptions can be made about order of hook execution. However, there are some limited guarantees
+about hook sequencing during install, upgrade and relation removal. 
+
+In normal operation, a unit will run at least the install, start, config-changed  and stop hooks over the course of its lifetime.
+
+### Install phase
+
+When a charm is first deployed, the following hook are executed in order before a charm reaches its operation phase:
+
+* storage-attached (if the unit has storage)
+* install
+* leader-elected (if the unit is a leader)
+* config-changed
+* start
+
+### Operation phase
+
+This phase is when a charm has completed its install operations and starts responding to events which correspond to interesting (relevant) changes to the Juju model.
+The behaviour is described in terms of hook kinds and is covered in subsequent sections.
+
+### Upgrade phase
+
+When a charm is upgraded, the `upgrade-charm` hook is usually followed by a `config-changed` hook.
+
+The `upgrade-charm` hook always runs once immediately after the charm directory
+contents have been changed by an unforced charm upgrade operation, and *may* do
+so after a forced upgrade; but will *not* be run after a forced upgrade from an
+existing error state. (Consequently, neither will the config-changed hook that
+would ordinarily follow the upgrade-charm.)
+
+
+### Tear down phase
+
+When an application is to be removed, the `stop` hook is the last hook to be run before the unit is destroyed.
+
+## Errors in hooks
+
+Hooks should ideally be idempotent, so that they can fail and be re-executed
+from scratch without trouble. As a hook author, you don't have complete control
+over the times your hook might be stopped: if the unit agent process is killed
+for any reason while running a hook, then when it recovers it will treat that
+hook as having failed -- just as if it had returned a non-zero exit code -- and
+request user intervention.
+
+It is unrealistic to expect great sophistication on the part of the average user,
+and as a charm author you should expect that users will attempt to re-execute
+failed hooks before attempting to investigate or understand the situation. You
+should therefore make every effort to ensure your hooks are idempotent when
+aborted and restarted.
+
```

Comment:
Addressing the charm author directly with "you" is a relic of when this doc was in the charm SDK. Let's rephrase.

"it is unrealistic to expect great sophistication on the part of the average user" is another relic from the charm SDK docs. Let's rephrase. The correct angle here is not whether the charm user _is_ sophisticated or not but rather that, on the Juju paradigm, they _don't have to be_.